src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexNodes.cs

```diff
@@ -0,0 +1,1643 @@
+        {
+            TypeToken = typeToken;
+        }
+
+        public RegexToken TypeToken { get; }
+    }
+
+    /// <summary>
+    /// A basic escape that just has \ and one additional character and needs no further information.
+    /// </summary>
+    internal sealed class RegexSimpleEscapeNode : RegexTypeEscapeNode
+    {
+        public RegexSimpleEscapeNode(RegexToken backslashToken, RegexToken typeToken)
+            : base(RegexKind.SimpleEscape, backslashToken, typeToken)
+        {
+            Debug.Assert(typeToken.Kind == RegexKind.TextToken);
+        }
+
+        internal override int ChildCount => 2;
+
+        internal override RegexNodeOrToken ChildAt(int index)
+        {
+            switch (index)
+            {
+            case 0: return BackslashToken;
+            case 1: return TypeToken;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// One of \b \B \A \G \z \Z
+    /// </summary>
+    internal sealed class RegexAnchorEscapeNode : RegexTypeEscapeNode
+    {
+        public RegexAnchorEscapeNode(RegexToken backslashToken, RegexToken typeToken)
+            : base(RegexKind.AnchorEscape, backslashToken, typeToken)
+        {
+        }
+
+        internal override int ChildCount => 2;
+
+        internal override RegexNodeOrToken ChildAt(int index)
+        {
+            switch (index)
+            {
+                case 0: return BackslashToken;
+                case 1: return TypeToken;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// One of \s \S \d \D \w \W
+    /// </summary>
+    internal sealed class RegexCharacterClassEscapeNode : RegexTypeEscapeNode
+    {
+        public RegexCharacterClassEscapeNode(RegexToken backslashToken, RegexToken typeToken)
+            : base(RegexKind.CharacterClassEscape, backslashToken, typeToken)
+        {
+        }
+
+        internal override int ChildCount => 2;
+
+        internal override RegexNodeOrToken ChildAt(int index)
+        {
+            switch (index)
+            {
+                case 0: return BackslashToken;
+                case 1: return TypeToken;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// ```\cX``` escape
+    /// </summary>
+    internal sealed class RegexControlEscapeNode : RegexTypeEscapeNode
+    {
+        public RegexControlEscapeNode(RegexToken backslashToken, RegexToken typeToken, RegexToken controlToken)
+            : base(RegexKind.ControlEscape, backslashToken, typeToken)
+        {
+            ControlToken = controlToken;
```

Comment:
Could assert `X`?

Reply:
ControlEscapes should be 'c' not 'x'.  But i get your point :)   I'm not sure how critical these types of asserts are.